internal/test/crds.go

```diff
@@ -0,0 +1,115 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gobuffalo/flect"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/pointer"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+)
+
+var (
+	bootstrapGroupVersion = schema.GroupVersion{Group: "bootstrap.cluster.x-k8s.io", Version: "v1alpha4"}
+
+	// BootstrapCRD is a generic boostrap CRD.
+	BootstrapCRD = generateCRD(bootstrapGroupVersion.WithKind("Bootstrap"))
+
+	// BootstrapTemplateCRD is a generic boostrap template CRD.
+	BootstrapTemplateCRD = generateCRD(bootstrapGroupVersion.WithKind("BootstrapTemplate"))
+
+	controlPlaneGroupVersion = schema.GroupVersion{Group: "controlplane.cluster.x-k8s.io", Version: "v1alpha4"}
+
+	// ControlPlaneCRD is a generic control plane CRD.
+	ControlPlaneCRD = generateCRD(controlPlaneGroupVersion.WithKind("ControlPlane"))
+
+	// ControlPlaneTemplateCRD is a generic control plane template CRD.
+	ControlPlaneTemplateCRD = generateCRD(controlPlaneGroupVersion.WithKind("ControlPlaneTemplate"))
+
+	infrastructureGroupVersion = schema.GroupVersion{Group: "infrastructure.cluster.x-k8s.io", Version: "v1alpha4"}
+
+	// InfrastructureMachineCRD is a generic infrastructure machine CRD.
+	InfrastructureMachineCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureMachine"))
+
+	// InfrastructureMachineTemplateCRD is a generic infrastructure machine template CRD.
+	InfrastructureMachineTemplateCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureMachineTemplate"))
+
+	// InfrastructureClusterCRD is a generic infrastructure machine CRD.
+	InfrastructureClusterCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureCluster"))
+
+	// InfrastructureClusterTemplateCRD is a generic infrastructure machine template CRD.
+	InfrastructureClusterTemplateCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureClusterTemplate"))
+
+	// InfrastructureRemediationCRD is a generic infrastructure remediation CRD.
+	InfrastructureRemediationCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureRemediation"))
+
+	// InfrastructureRemediationTemplateCRD is a generic infrastructure remediation template CRD.
+	InfrastructureRemediationTemplateCRD = generateCRD(infrastructureGroupVersion.WithKind("InfrastructureRemediationTemplate"))
+)
+
+func generateCRD(gvk schema.GroupVersionKind) *apiextensionsv1.CustomResourceDefinition {
```

Comment:
might for a future iteration, but PTAL to crd generation in clusterctl (it support both namespaced CRDs and global CRDs)

Reply:
Yup it has two exported wrappers for namespaced /cluster-wide CRDs and the unexported `fakeCRD` which doesn't set the scope. 

I went with the simplest implementation which matches what we currently need, which includes hard-coding scope as it's currently our only use case. I could already add either an additional parameter for scope or two wrapper funcs if we want. Also the func in clusterctl also adds a clusterctl label and has a paramater for versions.

Should I keep the current implementation or merge it with the one for clusterctl?

Reply:
ok, to keep current solution and iterate when we expand the scope for those CRDs, it was more and FYI than a request to change

Reply:
Got it :)